### ISSUE TITLE
NAT Traversal?

### ISSUE THREAD

**Issue by john3voltas** [git-lfs/git-lfs#2873]
Hello.
This is more a question than an issue.
We've been using git-lfs successfully on the LAN but we have guys working in customers premises in outsourcing.
These outsourcing guys have VPNs configured on their laptops but they have to keep connecting and disconnecting the VPN in order to keep their repos in sync, which is not feasible.
Can git-lfs be used in a NAT'ed environment? Or maybe using STUN/ICE or Magic Wormhole?
If so, could you please explain how we can achieve that?
Thanks in advance.
Cheers

**Comment by ttaylorr** [comment]
@john3voltas Hi, thanks for opening this issue and raising a great question. My understanding is that you'd like LFS to be able to work over NAT'ed networks or using STUN/ICE. I am not that familiar with these technologies, but can offer these pointers pertaining directly to Git LFS:

Git LFS makes HTTP to communicate with an LFS server (described in detail in: https://github.com/git-lfs/git-lfs/tree/master/docs/api). Beyond the initial HTTP call, anything that happens over the network (including NAT-ing) is transparent to Git LFS. So, if you can configure your Git LFS client to point at the correct address (by using `git config lfs.<remote>.url`), everything ought to work from Git LFS's perspective.

That all said, this repository offers support for issues pertaining to Git LFS specifically. Since I'm not familiar with the technologies that you listed above in depth currently, I don't think that I'll be able to help.

I'm going to close this for now, but if someone who is familiar with the above has insight that they would like to share, please do not hesitate to post it here.

**Comment by john3voltas** [comment]
@ttaylorr Hi, thanks for your reply.
Indeed I'm looking for a way to sync stuff (all kinds, small, medium and large files) using git (in this case git-lfs) with computers that are running off premises on customers' sites. Usually they will be behind a firewall that will be configured for NAT.
So this places three issues: resolving the remote IP address, sorting out which IP port the remote is listening on and punching a whole in the firewall. I don't even know the order in which this is done but the truth is this is needed if you are behind a firewall/NAT.

OTOH, the mechanism you mentioned is based on the premise that we have a centralized server/repository that sits behind our corporate firewall that is always listening on the same ip address and IP port, and the firewall allows any given public IP to test communication with *that* public IP address over *that* IP port.
The problem is, we don't control our corporate firewall and we don't have a server in our office. Actually we have a desktop on our office which we shutdown every end of the day because the building cuts the power everyday between 23h and 6h.
We're a small group of "young" consultants trying to startup and our very small office is located in a business center building that provides us with all the main services: internet, email boxes, VPN, phones, video conference, shared meeting rooms, cafeteria, you name it.
None of those are under our control so we can't configure a port forward on our external IP, one that we don't even know if it's fixed or if it's automatic.
That's why we thought about STUN, which we used to use with PBX'es and IP phones when we were all working from home. The phones used STUN to find out the IP and port of the IP PBX and that's how we managed to get telephony telephony working between our IP phones.

In our use case, it would really be very useful if git/git-lfs supported this feature but we understand that someone with more insight on the subject and with coding experience would have to jump in here and help developing such feature.
Hopefully someone like that will bump into this ticket and give us a help.
Many thanks for your support.
Cheers
